ships/0005-local-source-upload.md

```diff
@@ -0,0 +1,262 @@
+about the container image built, and instructions on how to pull and use the image.
+
+#### Watching the Build-Pod
+
+As mentioned before, Shipwright builds are based on a sequence of containers being executed on a
+single pod, the CLI needs to watch _(5)_ over the state changes, and react as soon as it happens. For
+example, the build-pod can be deleted, fail immediately, fail during execution, and such. Moreover,
+as we can observe on the sequence diagram, it also needs to wait for specific conditions before
+streaming data.
+
+During research and development, this requirement was translated into a [watch][ClientGoWatch]
+instantiated on the build-pod, and a straightforward mechanism to create an event-loop around pod
+[state changes][EventLoopExample].  Each change of state will trigger the call of the informed
+closures, and allows wiring [events and reactions in a practical manner][EventLoopUsage].
+
+### Upload Mechanism
+
+This EP describes the direct data streaming to the Kubernetes cluster, directly to the POD running
+the actual build process, while we already have plans to support data upload towards a
+[container-registry][LocalSourceCodeEP] alternatively.
+
+#### Redirecting Standard Input
+
+The data upload mechanism is the same employed on <code>[kubectl cp][KubectlCP]</code>, which is
+redirecting the standard-input (<code>STDIN</code>) from the client to the container, inside the
+desired pod.
+
+Analogous to `kubectl cp`, we will be using `tar` to stream data from the client, while
+<code>[kubectl exec][KubectlExec]</code> on the pod keeps a <code>tar</code> process waiting for
+data on <code>STDIN</code>.
+
+#### Shipwright Build Controller
+
+The Shipwright Build Controller works based on [Tekton Workspaces][TektonWorkspaces], and therefore,
+it already defines the /workspace/source mount point for all containers in the build-pod. Tekton also
+adds it's own helper to ["git clone" the repository][ControllerGitClone] in place.
+
+With that said, we can simply stop Shipwright Build-Controller from issuing the "`git clone`", and
+instead, add a [simple helper application][ControllerWaiter] that's able to hold the build-pod
+execution while the data is transferred.
+
+To accomplish this the Build-Controller needs to be able to detect it should replace the "`git
+clone`" procedure by waiting for the upload helper instead. The changes done during research and
+development are [here][ControllerWaiterUsage].
+
+##### BuildRun Changes
+
+The proposed API changes are the following:
+
+```yml
+---
+apiVersion: shipwright.io/v1alpha1
+kind: BuildRun
+metadata: {}
+spec:
+  sources:
+    - name: src
+      type: LocalCopy
+```
+
+The Build Controller will inspect for a type `LocalCopy` on the "`sources"` slice, and therefore
+instead of adding the git clone step, it will instead wait for the user to upload.
+
+The `sources` slice is part of [Remote Artifacts (EP)][RemoteArtifactsEP] and has been worked on, so
```

Comment:
we need to put in that EP ( _remote artifacts_ ) the final state of sources before we merge this one. As this is introducing a new type `LocalCopy`. I´m adding this to the next community meeting, see https://github.com/shipwright-io/community/issues/15#issuecomment-874708101 .

The Remote Artifacts EP requires at least the following enhancements:
- proposal on the bundle type (local source code EP)
- proposal on the LocalCopy type (this EP)

Reply:
Sure! I understand the mix with Remote Artifacts due the initial text in this enhancement-proposal, so I'm now updating it it allow both EP implementations to go stand-alone. Here we have a API suggestion for when the Remote Artifact EP implementation comes along, and a method to not require API changes at all.

On amending the document I'm inclined to annotations instead of a full blown API change. Let me know your thoughts.